slides/12_Remix.html

```diff
@@ -24,6 +24,727 @@
+                  returned data
+                </li>
+                <li>request is the Fetch API Request interface</li>
+                <li>
+                  params has route parameters (use request for query params)
+                </li>
+                <li>On initial request runs internally on server</li>
+                <li>
+                  When navigating to the page within SPA, performs background
+                  request
+                </li>
+              </ul>
+            </aside>
+          </section>
+          <section data-auto-animate>
+            <h3>Modifying the HTTP response</h3>
+          </section>
+          <section data-auto-animate>
+            <h3>Modifying the HTTP response</h3>
+
+            <p>Return a <span class="code">Response</span> object…</p>
+            <pre><code class="hljs tsx" data-trim data-line-numbers="4-6" data-id="code-1"><script type="text/template">
+              export async function loader({ request, params, context }: LoaderArgs) {
+                …
+
+                if (notFound) {
+                  return new Response(null, { status: 404, statusText: "Not found" });
+                }
+
+                return json(data);
+              }
+            </script></code></pre>
+            <aside class="notes">
+              <ul>
+                <li>Also useful for 400 (Bad Request) or modifying headers</li>
+              </ul>
+            </aside>
+          </section>
+          <section data-auto-animate>
+            <h3>Modifying the HTTP response</h3>
+            <p>…or <em>throw</em> it!</p>
+            <pre><code class="hljs tsx" data-trim data-line-numbers="9,17|12-14|4-6" data-id="code-2"><script type="text/template">
+              export async function loader({ request, params, context }: LoaderArgs) {
+                …
+
+                const post = await getBlogPost(params.slug);
+
+                return json(data);
+              }
+
+              export async function getBlogPost(slug: string) {
+                …
+
+                if (notFound) {
+                  throw new Response(null, { status: 404, statusText: "Not found" });
+                }
+
+                return post;
+              }
+            </script></code></pre>
+            <aside class="notes">
+              <ul>
+                <li>This way, we don't need an if in the loader</li>
+                <li>
+                  Also doesn't break inference of the loader data like an if
+                  would
+                </li>
+              </ul>
+            </aside>
+          </section>
+        </section>
+        <section>
+          <h2>Exercise</h2>
+          <ul>
+            <li class="fragment">
+              Load the data for the list and detail pages from the Poke API
+            </li>
+            <li class="fragment">
+              Show the Not Found page if an unknown pokemon name is passed in
+              the URL
+            </li>
```

Comment:
**Suggestion:** What about a stretch goal where we add logic to the loader? I.e. strip properties from the result object, that we don't need in the UI, so on a page navigation only the desired data is sent to the frontend?

Reply:
Good idea, it shows the power of the Remix loaders as opposed to directly loading from the 3rd party API.